torch/csrc/jit/passes/peephole_list_idioms.cpp

```diff
@@ -21,17 +22,365 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+              true_block_throws ? n->blocks().at(1) : n->blocks().at(0);
+          for (size_t i = 0; i < if_n.outputs().size(); ++i) {
+            if (boolean_value_refinements_.count(
+                    non_throwing_block->outputs().at(i))) {
+              boolean_value_refinements_[n->outputs().at(i)] =
+                  boolean_value_refinements_[non_throwing_block->outputs().at(
+                      i)];
+            }
+          }
+          continue;
+        }
+
+        for (size_t i = 0; i < if_n.outputs().size(); ++i) {
+          if (!(if_n.outputs().at(i)->type() == BoolType::get())) {
+            continue;
+          }
+          Value* true_v = if_n.thenOutputs().at(i);
+          Value* false_v = if_n.elseOutputs().at(i);
+
+          if (!boolean_value_refinements_.count(true_v) &&
+              !boolean_value_refinements_.count(false_v) &&
+              !constant_as<bool>(true_v) && !constant_as<bool>(false_v)) {
+            continue;
+          }
+
+          // if either block has a constant bool output, e.g. `true` on the
+          // true block, then for the `false` value we can take the false
+          // refinements present on the false block and from the other block
+          // output value bc if the output is false it had to have come from the
+          // false block. if len(a) == 5:
+          //     x = len(b) == 4
+          // else:
+          //     x = False
+          // if x is true, then we know both len(a) == 5 and len(b) == 4
+          //
+          // if neither block has a constant bool value, we just take the
+          // intersection of the refinements from boolean outputs.
+          // if cond:
+          //    x = len(a) == 4 and len(b) == 5
+          // else:
+          //    x = len(a) == 4
+          // here, we know if x is true, then len(a) == 4, but not len(b)
+          // == 5, because that refinement is not present in the true block.
+          // TODO: could also take intersection of refinements present in
+          // both blocks, but it's not a real use case.
+
+          // boolean_value_refinements_[value] is safe to access because
+          // BooleanRefinementMapping has a default constructor
+
+          BooleanRefinementMapping out;
+          if (auto maybe_bool = constant_as<bool>(true_v)) {
+            if (*maybe_bool) {
+              out = BooleanRefinementMapping::FalseRefinements(unionRefinements(
+                  boolean_value_refinements_[false_v].false_refine(),
+                  false_block_refinements));
+            } else {
+              out = BooleanRefinementMapping::TrueRefinements(unionRefinements(
+                  boolean_value_refinements_[false_v].true_refine(),
+                  false_block_refinements));
+            }
+          } else if (auto maybe_bool = constant_as<bool>(false_v)) {
+            if (*maybe_bool) {
+              out = BooleanRefinementMapping::FalseRefinements(unionRefinements(
```

Comment:
Nit: would be helpful to add an example to each of these cases to make it easier trace the logic.